python/pyarrow/error.pyx

```diff
@@ -19,13 +19,52 @@ from pyarrow.includes.libarrow cimport CStatus
 from pyarrow.includes.common cimport c_string
 from pyarrow.compat import frombytes
 
+
 class ArrowException(Exception):
     pass
 
+
+class ArrowInvalid(ValueError, ArrowException):
```

Comment:
my main concern was to NOT have the arrow exception hierarchy just be based on ``Exception``. Being able to catch ``ValueError``, ``TypeError``, ``KeyError``, ``IndexError`` (subclass) as errors is very pythonic.

As for the naminghaving ``ArrowTypeError(ArrowException, TypeError)``. I am not how useful that actually is in practice. I would say if you have a need to differentiate between *two* (or more ``TypeError`` s), then having a custom exception is useful (as opposed to having to check the actual error message).

So fine with this PR (though maybe raise the standard python exceptions, unless you have a real need for them).